Problem: unnecessary contract account updates

yihuang opened the pull request:
When a contract's state is changed, statedb marks the contract address as dirty, then when commit, the account object get updated, although the account object has nothing changed.

Solution:
- compare the account fields before set

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

Closes: #XXX

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

yihuang commented on CHANGELOG.md:
```diff
@@ -95,6 +95,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 - (precompile) [#383](https://github.com/crypto-org-chain/ethermint/pull/383) Allow init precompiled contract with ctx.
 - (evm) [#393](https://github.com/crypto-org-chain/ethermint/pull/393) Cleanup ApplyMessageWithConfig interface.
 - (cmd) [#399](https://github.com/crypto-org-chain/ethermint/pull/399) Add pruning and snapshot cmd.
+- (statedb) [#]() Only set account when it's actually changed.
```
```suggestion
- (statedb) [#410](https://github.com/crypto-org-chain/ethermint/pull/410) Only set account when it's actually changed.
```

yihuang commented:
a better solution @mmsqe

mmsqe commented on x/evm/statedb/state_object.go:
```diff
@@ -88,14 +90,25 @@ func newObject(db *StateDB, address common.Address, account Account) *stateObjec
 	}
 }
 
+// codeDirty returns wether the code is modified
```
```suggestion
// codeDirty returns whether the codeHash is modified
```

Commit:
Problem: unnesserary contract account updates

When a contract's state is changed, the contract address is marked as dirty, and the account object is set as well,
although the account object has nothing changed.

Soluton:
- compare the account fields before set

Update CHANGELOG.md

Signed-off-by: yihuang <huang@crypto.com>

Update CHANGELOG.md

Signed-off-by: yihuang <huang@crypto.com>

solve the problem at higher level

Update x/evm/statedb/state_object.go

Co-authored-by: mmsqe <mavis@crypto.com>
Signed-off-by: yihuang <huang@crypto.com>

mmsqe commented on x/evm/statedb/statedb.go:
```diff
@@ -612,11 +612,14 @@ func (s *StateDB) Commit() error {
 			}
-			if err := s.keeper.SetAccount(s.ctx, obj.Address(), obj.account); err != nil {
-				return errorsmod.Wrap(err, "failed to set account")
+			if codeDirty || obj.nonceDirty() {
```
do we miss check obj.empty() here

yihuang replied:
what empty?

yihuang replied:
what do you mean? delete empty accounts?

mmsqe replied:
I mean we also need SetAccount for new account?

yihuang replied:
i see, we don't have to, but if not,it'll be consensus-breaking.

yihuang replied:
done

Commit:
save the new account
